Add config for setting the ACL type for S3 uploads

kash159 opened the pull request:
The purpose of this diff is to enable specifying an ACL Policy when uploading objects into S3 via the presto S3 File System. Our main motivation is that we have multiple accounts for our organization within AWS and we want to = enable cross account writes without breaking backwards compatibility. To do this, we need to set the proper canned ACL. (bucket owner full control in our case) This settings allows us to specify ACL.

kash159 commented:
@nezihyigitbasi @electrum @zhenxiao

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/HiveS3Config.java:
```diff
@@ -56,6 +56,7 @@
     private DataSize s3MultipartMinPartSize = new DataSize(5, MEGABYTE);
     private boolean pinS3ClientToCurrentRegion;
     private String s3UserAgentPrefix = "";
+    private PrestoS3AclType s3AclType = PrestoS3AclType.Private;
```
Is `Private` also the default acl that aws sdk uses when we don't specify it? We should confirm that before making this the default.

kash159 replied:
According to https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/CannedAccessControlList.html: 

public static final CannedAccessControlList Private
Specifies the owner is granted Permission.FullControl. No one else has access rights.
This is the default access control policy for any new buckets or objects.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/HiveS3Config.java:
```diff
@@ -374,4 +375,18 @@ public HiveS3Config setS3UserAgentPrefix(String s3UserAgentPrefix)
+        return s3AclType;
+    }
+
+    @Config("hive.s3.acl.type")
```
Since this is only affecting the upload/write path, we can find a better name to reflect it. Maybe `hive.s3.upload-acl-type`.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3AclType.java:
```diff
@@ -0,0 +1,41 @@
+    PublicRead(CannedAccessControlList.PublicRead),
+    PublicReadWrite(CannedAccessControlList.PublicReadWrite);
+
+    public CannedAccessControlList getCannedACL()
```
this can be package private

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3AclType.java:
```diff
@@ -0,0 +1,41 @@
+import com.amazonaws.services.s3.model.CannedAccessControlList;
+
+@SuppressWarnings("EnumeratedConstantNamingConvention")
+public enum PrestoS3AclType
```
this enum can be package private

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java:
```diff
@@ -1056,6 +1062,8 @@ private void uploadObject()
                     }
                 }
 
+                request.withCannedAcl(this.aclType);
```
remove `this.` to be consistent with other field accesses here

nezihyigitbasi commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -531,4 +533,39 @@ public AWSCredentials getCredentials()
+            String expectedBucketName = "test-bucket";
+            fs.initialize(new URI("s3n://" + expectedBucketName + "/"), config);
+            fs.setS3Client(s3);
+            FSDataOutputStream stream = fs.create(new Path("s3n://test-bucket/test"));
```
You can also do

```
try (FSDataOutputStream stream = fs.create(new Path("s3n://test-bucket/test")))
{
    assertEquals(CannedAccessControlList.Private, s3.getAcl());
}
```

nezihyigitbasi commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -531,4 +533,39 @@ public AWSCredentials getCredentials()
+            fs.initialize(new URI("s3n://" + expectedBucketName + "/"), config);
+            fs.setS3Client(s3);
+            FSDataOutputStream stream = fs.create(new Path("s3n://test-bucket/test"));
+            stream.close();
```
ditto

nezihyigitbasi commented:
@electrum  thoughts?

kash159 commented:
@nezihyigitbasi  addressed your comments. Let me know if you have any concerns.

According to the doc, default looks to be Private. Let me know if that is okay. We can also only set this if only if the option if specified. Let me know which way you prefer.

nezihyigitbasi commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -531,4 +533,41 @@ public AWSCredentials getCredentials()
+            String expectedBucketName = "test-bucket";
+            fs.initialize(new URI("s3n://" + expectedBucketName + "/"), config);
+            fs.setS3Client(s3);
+            try (FSDataOutputStream stream = fs.create(new Path("s3n://test-bucket/test"))) {
```
```
            try(FSDataOutputStream stream = fs.create(new Path("s3n://test-bucket/test"))) {
                // initiate the upload by creating a stream & closing it immediately
            }
            assertEquals(s3.getAcl(), CannedAccessControlList.Private);
```

nezihyigitbasi commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -531,4 +533,41 @@ public AWSCredentials getCredentials()
+            String expectedBucketName = "test-bucket";
+            fs.initialize(new URI("s3n://" + expectedBucketName + "/"), config);
+            fs.setS3Client(s3);
+            try (FSDataOutputStream stream = fs.create(new Path("s3n://test-bucket/test"))) {
```
```
            try(FSDataOutputStream stream = fs.create(new Path("s3n://test-bucket/test"))) {
                // initiate the upload by creating a stream & closing it immediately
            }
            assertEquals(s3.getAcl(), CannedAccessControlList.BucketOwnerFullControl);
```

kash159 commented:
This should be good to go @nezihyigitbasi

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java:
```diff
@@ -998,13 +1004,13 @@ public PrestoS3OutputStream(
                     .withMinimumUploadPartSize(multiPartUploadMinPartSize)
                     .withMultipartUploadThreshold(multiPartUploadMinFileSize).build();
 
+            this.aclType = requireNonNull(aclType.getCannedACL(), "aclType is null");
```
```suggestion
            this.aclType = requireNonNull(aclType, "aclType is null");
```

nezihyigitbasi commented:
Please update the commit message title as `Add config for setting the ACL type for S3 uploads`

nezihyigitbasi commented:
If you rebase against the current master & address my two comments I am going to get this in after the release goes out.

kash159 commented:
done. ty for the help!

kash159 commented:
Can we merge this @nezihyigitbasi now that .213 is out.

nezihyigitbasi commented:
Merged with minor changes (doc change, whitespace change, and an enum naming change), thanks @kash159!
